File: backend/firestore_service.js
const admin = require('firebase-admin');

/**
 * Firestore service for backend operations
 *
 * This service handles Firestore operations on the backend using Firebase Admin SDK.
 * It replaces the SQLite database with cloud-based Firestore for better scalability,
 * real-time updates, and seamless integration with the frontend.
 */
class FirestoreService {
  constructor() {
    this.isInitialized = false;
    this.db = null;
  }

  /**
   * Initialize Firestore service with Firebase Admin SDK
   */
  async initialize() {
    try {
      if (this.isInitialized) {
        console.log('🔥 Firestore Service already initialized');
        return;
      }

      // Use the same Firebase app instance from FCM service if available
      let app;
      try {
        app = admin.app(); // Get default app if it exists
      } catch (error) {
        // Initialize if no app exists
        try {
          const serviceAccount = require('./firebase-service-account.json');
          app = admin.initializeApp({
            credential: admin.credential.cert(serviceAccount),
            projectId: serviceAccount.project_id,
          });
          console.log('🔥 Firebase Admin initialized with service account');
        } catch (fileError) {
          app = admin.initializeApp({
            projectId: process.env.FIREBASE_PROJECT_ID || 'your-project-id',
          });
          console.log('🔥 Firebase Admin initialized with default credentials');
        }
      }

      this.db = admin.firestore();

      // Configure Firestore settings
      this.db.settings({
        ignoreUndefinedProperties: true,
      });

      this.isInitialized = true;
      console.log('✅ Firestore Service initialized successfully');

    } catch (error) {
      console.error('❌ Error initializing Firestore Service:', error.message);
      throw error;
    }
  }

  /**
   * Insert a new report into Firestore
   */
  async insertReport(report) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log(`📄 Inserting report into Firestore: ${report.id}`);

      // Transform the report data to match Firestore structure
      const firestoreData = {
        id: report.id,
        foto_url: report.image_path || '',
        ubicacion: report.location ? `${report.location.latitude}, ${report.location.longitude}` : '',
        clasificacion: report.classification || '',
        estado: report.status || 'received',
        prioridad: 'Media', // Default priority
        tipo_residuo: report.classification || '',
        location: {
          latitude: report.location?.latitude || 0,
          longitude: report.location?.longitude || 0,
          accuracy: report.location?.accuracy || 0,
        },
        created_at: admin.firestore.Timestamp.fromDate(new Date(report.timestamp || new Date())),
        updated_at: admin.firestore.Timestamp.now(),
        device_info: report.device_info || '',
        user_id: report.user_id || null,
      };

      // Use the report ID as the document ID
      await this.db.collection('reports').doc(report.id).set(firestoreData);

      console.log(`✅ Report ${report.id} inserted into Firestore successfully`);
      return { success: true, id: report.id };

    } catch (error) {
      console.error('❌ Error inserting report into Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Get a report by ID from Firestore
   */
  async getReport(reportId) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log(`📖 Getting report from Firestore: ${reportId}`);

      const doc = await this.db.collection('reports').doc(reportId).get();

      if (!doc.exists) {
        console.log(`⚠️ Report not found: ${reportId}`);
        return null;
      }

      const data = doc.data();

      // Transform Firestore data back to expected format
      const report = {
        id: doc.id,
        timestamp: data.created_at.toDate().toISOString(),
        location: {
          latitude: data.location.latitude,
          longitude: data.location.longitude,
          accuracy: data.location.accuracy,
        },
        classification: data.clasificacion,
        device_info: data.device_info,
        image_path: data.foto_url,
        status: data.estado,
        created_at: data.created_at.toDate().toISOString(),
        updated_at: data.updated_at.toDate().toISOString(),
      };

      console.log(`✅ Report retrieved from Firestore: ${reportId}`);
      return report;

    } catch (error) {
      console.error('❌ Error getting report from Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Get all reports with pagination
   */
  async getAllReports(limit = 100, offset = 0) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log(`📋 Getting all reports from Firestore (limit: ${limit}, offset: ${offset})`);

      // Use Firestore pagination with startAfter for better performance
      let query = this.db
        .collection('reports')
        .orderBy('created_at', 'desc')
        .limit(limit);

      // For offset, we need to get the document to start after
      if (offset > 0) {
        const offsetDoc = await this.db
          .collection('reports')
          .orderBy('created_at', 'desc')
          .offset(offset)
          .limit(1)
          .get();

        if (!offsetDoc.empty) {
          query = query.startAfter(offsetDoc.docs[0]);
        }
      }

      const snapshot = await query.get();

      const reports = snapshot.docs.map(doc => {
        const data = doc.data();
        return {
          id: doc.id,
          timestamp: data.created_at.toDate().toISOString(),
          location: {
            latitude: data.location.latitude,
            longitude: data.location.longitude,
            accuracy: data.location.accuracy,
          },
          classification: data.clasificacion,
          device_info: data.device_info,
          image_path: data.foto_url,
          status: data.estado,
          created_at: data.created_at.toDate().toISOString(),
          updated_at: data.updated_at.toDate().toISOString(),
        };
      });

      console.log(`✅ Retrieved ${reports.length} reports from Firestore`);
      return reports;

    } catch (error) {
      console.error('❌ Error getting all reports from Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Update report status in Firestore
   */
  async updateReportStatus(reportId, newStatus, timestamp = null) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log(`🔄 Updating report status in Firestore: ${reportId} -> ${newStatus}`);

      const updateData = {
        estado: newStatus,
        updated_at: admin.firestore.Timestamp.now(),
      };

      if (timestamp) {
        updateData.timestamp = admin.firestore.Timestamp.fromDate(new Date(timestamp));
      }

      const docRef = this.db.collection('reports').doc(reportId);

      // Check if document exists
      const doc = await docRef.get();
      if (!doc.exists) {
        throw new Error(`Report with ID ${reportId} not found`);
      }

      await docRef.update(updateData);

      console.log(`✅ Report status updated in Firestore: ${reportId} -> ${newStatus}`);
      return { success: true, changes: 1 };

    } catch (error) {
      console.error('❌ Error updating report status in Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Get statistics from Firestore
   */
  async getStats() {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log('📊 Getting statistics from Firestore...');

      // Get total reports count
      const totalQuery = await this.db.collection('reports').count().get();
      const totalReports = totalQuery.data().count;

      // Get reports by classification
      const classificationStats = {};
      const snapshot = await this.db.collection('reports').get();

      snapshot.docs.forEach(doc => {
        const data = doc.data();
        const classification = data.clasificacion || 'Unknown';
        classificationStats[classification] = (classificationStats[classification] || 0) + 1;
      });

      // Get reports by status
      const statusStats = {};
      snapshot.docs.forEach(doc => {
        const data = doc.data();
        const status = data.estado || 'Unknown';
        statusStats[status] = (statusStats[status] || 0) + 1;
      });

      const stats = {
        total_reports: totalReports,
        classifications: classificationStats,
        status_breakdown: statusStats,
        last_updated: new Date().toISOString(),
      };

      console.log('✅ Statistics retrieved from Firestore:', stats);
      return stats;

    } catch (error) {
      console.error('❌ Error getting statistics from Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Delete a report from Firestore
   */
  async deleteReport(reportId) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log(`🗑️ Deleting report from Firestore: ${reportId}`);

      const docRef = this.db.collection('reports').doc(reportId);

      // Check if document exists
      const doc = await docRef.get();
      if (!doc.exists) {
        throw new Error(`Report with ID ${reportId} not found`);
      }

      await docRef.delete();

      console.log(`✅ Report deleted from Firestore: ${reportId}`);
      return { success: true };

    } catch (error) {
      console.error('❌ Error deleting report from Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Get reports by status
   */
  async getReportsByStatus(status, limit = 50) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log(`📋 Getting reports by status from Firestore: ${status}`);

      const snapshot = await this.db
        .collection('reports')
        .where('estado', '==', status)
        .orderBy('created_at', 'desc')
        .limit(limit)
        .get();

      const reports = snapshot.docs.map(doc => {
        const data = doc.data();
        return {
          id: doc.id,
          timestamp: data.created_at.toDate().toISOString(),
          location: {
            latitude: data.location.latitude,
            longitude: data.location.longitude,
            accuracy: data.location.accuracy,
          },
          classification: data.clasificacion,
          device_info: data.device_info,
          image_path: data.foto_url,
          status: data.estado,
          created_at: data.created_at.toDate().toISOString(),
        };
      });

      console.log(`✅ Retrieved ${reports.length} reports with status ${status}`);
      return reports;

    } catch (error) {
      console.error('❌ Error getting reports by status from Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Register FCM token in Firestore
   */
  async registerFCMToken(token, userId, metadata = {}) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log(`📱 Registering FCM token in Firestore: ${token.substring(0, 20)}...`);

      const tokenData = {
        token,
        user_id: userId || 'anonymous',
        platform: metadata.platform || 'unknown',
        app_version: metadata.app_version || '1.0.0',
        device_info: metadata.device_info || 'unknown',
        created_at: admin.firestore.Timestamp.now(),
        updated_at: admin.firestore.Timestamp.now(),
        last_used: admin.firestore.Timestamp.now(),
      };

      // Use token as document ID or generate one
      const docId = userId || token.substring(0, 20);
      await this.db.collection('fcm_tokens').doc(docId).set(tokenData, { merge: true });

      console.log(`✅ FCM token registered in Firestore: ${docId}`);
      return { success: true };

    } catch (error) {
      console.error('❌ Error registering FCM token in Firestore:', error.message);
      throw error;
    }
  }

  /**
   * Get FCM tokens from Firestore
   */
  async getFCMTokens(userId = null) {
    try {
      if (!this.isInitialized) {
        throw new Error('Firestore service not initialized');
      }

      console.log('📱 Getting FCM tokens from Firestore...');

      let query = this.db.collection('fcm_tokens');

      if (userId) {
        query = query.where('user_id', '==', userId);
      }

      const snapshot = await query.get();
      const tokens = snapshot.docs
        .map(doc => doc.data().token)
        .filter(token => token);

      console.log(`✅ Retrieved ${tokens.length} FCM tokens from Firestore`);
      return tokens;

    } catch (error) {
      console.error('❌ Error getting FCM tokens from Firestore:', error.message);
      return [];
    }
  }

  /**
   * Close Firestore connection (cleanup)
   */
  close() {
    if (this.isInitialized) {
      console.log('🔥 Closing Firestore connection...');
      // Firestore connections are managed by Firebase Admin SDK
      // No explicit close needed
      this.isInitialized = false;
    }
  }
}

// Export singleton instance
module.exports = new FirestoreService();

File: PaginaWeb_P2/reports/firestore_service.py
"""
Firestore Service for Django Integration

This service provides a bridge between Django and Firebase Firestore,
allowing the Django web application to read and write reports from/to
the same Firestore database used by the mobile app and backend API.
"""

import firebase_admin
from firebase_admin import credentials, firestore
from datetime import datetime
from django.conf import settings
import os


class FirestoreService:
    """Service class to handle Firestore operations for Django"""

    _instance = None
    _initialized = False

    def __new__(cls):
        """Singleton pattern to ensure only one Firebase connection"""
        if cls._instance is None:
            cls._instance = super(FirestoreService, cls).__new__(cls)
        return cls._instance

    def __init__(self):
        """Initialize Firebase Admin SDK"""
        if not FirestoreService._initialized:
            try:
                # Get credentials path from settings or use default
                cred_path = getattr(settings, 'FIREBASE_CREDENTIALS',
                                   os.path.join(settings.BASE_DIR, 'firebase-service-account.json'))

                # Initialize Firebase Admin
                if not firebase_admin._apps:
                    cred = credentials.Certificate(cred_path)
                    firebase_admin.initialize_app(cred)
                    print('🔥 Firebase Admin SDK initialized successfully')

                self.db = firestore.client()
                FirestoreService._initialized = True
                print('✅ Firestore Service ready')

            except Exception as e:
                print(f'❌ Error initializing Firestore Service: {e}')
                raise

    # ==================== READ OPERATIONS ====================

    def get_all_reports(self, limit=None, order_by='created_at', descending=True):
        """
        Get all reports from Firestore

        Args:
            limit: Maximum number of reports to retrieve
            order_by: Field to order by (default: created_at)
            descending: Order descending (default: True)

        Returns:
            List of report dictionaries
        """
        try:
            print(f'📋 Fetching all reports from Firestore (limit: {limit})')

            # Build query
            query = self.db.collection('reports').order_by(
                order_by,
                direction=firestore.Query.DESCENDING if descending else firestore.Query.ASCENDING
            )

            if limit:
                query = query.limit(limit)

            # Execute query
            docs = query.stream()

            # Convert to Django-compatible format
            reports = [self._firestore_to_django(doc) for doc in docs]

            print(f'✅ Retrieved {len(reports)} reports from Firestore')
            return reports

        except Exception as e:
            print(f'❌ Error fetching reports from Firestore: {e}')
            return []

    def get_report_by_id(self, report_id):
        """
        Get a single report by ID

        Args:
            report_id: The report ID

        Returns:
            Report dictionary or None
        """
        try:
            print(f'📖 Fetching report {report_id} from Firestore')

            doc = self.db.collection('reports').document(report_id).get()

            if doc.exists:
                report = self._firestore_to_django(doc)
                print(f'✅ Report {report_id} retrieved')
                return report
            else:
                print(f'⚠️ Report {report_id} not found in Firestore')
                return None

        except Exception as e:
            print(f'❌ Error fetching report {report_id}: {e}')
            return None

    def get_reports_by_status(self, status, limit=None):
        """
        Get reports filtered by status

        Args:
            status: Status to filter by (e.g., 'received', 'in_progress', 'completed')
            limit: Maximum number of reports

        Returns:
            List of report dictionaries
        """
        try:
            print(f'📋 Fetching reports with status: {status}')

            query = self.db.collection('reports').where('estado', '==', status)
            query = query.order_by('created_at', direction=firestore.Query.DESCENDING)

            if limit:
                query = query.limit(limit)

            docs = query.stream()
            reports = [self._firestore_to_django(doc) for doc in docs]

            print(f'✅ Retrieved {len(reports)} reports with status {status}')
            return reports

        except Exception as e:
            print(f'❌ Error fetching reports by status: {e}')
            return []

    def get_reports_by_classification(self, clasificacion, limit=None):
        """
        Get reports filtered by classification (tipo_residuo)

        Args:
            clasificacion: Classification to filter by
            limit: Maximum number of reports

        Returns:
            List of report dictionaries
        """
        try:
            query = self.db.collection('reports').where('clasificacion', '==', clasificacion)
            query = query.order_by('created_at', direction=firestore.Query.DESCENDING)

            if limit:
                query = query.limit(limit)

            docs = query.stream()
            reports = [self._firestore_to_django(doc) for doc in docs]

            return reports

        except Exception as e:
            print(f'❌ Error fetching reports by classification: {e}')
            return []

    def get_reports_by_priority(self, prioridad, limit=None):
        """
        Get reports filtered by priority

        Args:
            prioridad: Priority to filter by
            limit: Maximum number of reports

        Returns:
            List of report dictionaries
        """
        try:
            query = self.db.collection('reports').where('prioridad', '==', prioridad)
            query = query.order_by('created_at', direction=firestore.Query.DESCENDING)

            if limit:
                query = query.limit(limit)

            docs = query.stream()
            reports = [self._firestore_to_django(doc) for doc in docs]

            return reports

        except Exception as e:
            print(f'❌ Error fetching reports by priority: {e}')
            return []

    # ==================== WRITE OPERATIONS ====================

    def update_report_status(self, report_id, new_status):
        """
        Update the status of a report

        Args:
            report_id: The report ID
            new_status: New status value

        Returns:
            Boolean indicating success
        """
        try:
            print(f'🔄 Updating report {report_id} status to: {new_status}')

            doc_ref = self.db.collection('reports').document(report_id)
            doc_ref.update({
                'estado': new_status,
                'updated_at': firestore.SERVER_TIMESTAMP
            })

            print(f'✅ Report {report_id} status updated to {new_status}')
            return True

        except Exception as e:
            print(f'❌ Error updating report status: {e}')
            return False

    def update_report(self, report_id, updates):
        """
        Update a report with custom fields

        Args:
            report_id: The report ID
            updates: Dictionary with fields to update

        Returns:
            Boolean indicating success
        """
        try:
            print(f'🔄 Updating report {report_id} with {len(updates)} fields')

            # Always add updated_at timestamp
            updates['updated_at'] = firestore.SERVER_TIMESTAMP

            doc_ref = self.db.collection('reports').document(report_id)
            doc_ref.update(updates)

            print(f'✅ Report {report_id} updated successfully')
            return True

        except Exception as e:
            print(f'❌ Error updating report: {e}')
            return False

    def assign_report_to_user(self, report_id, user_id, user_name):
        """
        Assign a report to a user (encargado)

        Args:
            report_id: The report ID
            user_id: The user ID to assign to
            user_name: The user's full name

        Returns:
            Boolean indicating success
        """
        try:
            updates = {
                'assigned_to': user_id,
                'assigned_to_name': user_name,
                'assigned_at': firestore.SERVER_TIMESTAMP,
                'estado': 'asignado'
            }

            return self.update_report(report_id, updates)

        except Exception as e:
            print(f'❌ Error assigning report: {e}')
            return False

    # ==================== STATISTICS ====================

    def get_stats(self):
        """
        Get statistics about reports

        Returns:
            Dictionary with statistics
        """
        try:
            print('📊 Calculating Firestore statistics...')

            # Get all reports
            all_reports = self.get_all_reports()

            # Calculate stats
            stats = {
                'total_reports': len(all_reports),
                'by_status': {},
                'by_classification': {},
                'by_priority': {},
            }

            # Count by status
            for report in all_reports:
                status = report.get('estado', 'unknown')
                stats['by_status'][status] = stats['by_status'].get(status, 0) + 1

            # Count by classification
            for report in all_reports:
                clasificacion = report.get('tipo_residuo', 'unknown')
                stats['by_classification'][clasificacion] = stats['by_classification'].get(clasificacion, 0) + 1

            # Count by priority
            for report in all_reports:
                prioridad = report.get('prioridad', 'unknown')
                stats['by_priority'][prioridad] = stats['by_priority'].get(prioridad, 0) + 1

            print(f'✅ Statistics calculated: {stats["total_reports"]} total reports')
            return stats

        except Exception as e:
            print(f'❌ Error calculating statistics: {e}')
            return {}

    # ==================== HELPER METHODS ====================

    def _firestore_to_django(self, doc):
        """
        Convert Firestore document to Django-compatible dictionary

        Args:
            doc: Firestore document snapshot

        Returns:
            Dictionary compatible with Django templates and views
        """
        data = doc.to_dict()

        # Extract location data
        location = data.get('location', {})
        latitud = location.get('latitude', 0)
        longitud = location.get('longitude', 0)

        # Extract timestamps
        created_at = data.get('created_at')
        if created_at and hasattr(created_at, 'timestamp'):
            fecha_reporte = datetime.fromtimestamp(created_at.timestamp())
        else:
            fecha_reporte = datetime.now()

        # Map Firestore fields to Django model fields
        return {
            'id': doc.id,
            'tipo_residuo': self._map_clasificacion_to_django(data.get('clasificacion', '')),
            'tipo_residuo_display': data.get('clasificacion', 'Desconocido'),
            'descripcion': data.get('descripcion', ''),
            'foto': None,  # Not used with Firestore
            'foto_url': data.get('foto_url', ''),
            'latitud': latitud,
            'longitud': longitud,
            'direccion': data.get('ubicacion', ''),
            'estado': self._map_estado_to_django(data.get('estado', 'received')),
            'estado_display': data.get('estado', 'Recibido'),
            'prioridad': self._map_prioridad_to_django(data.get('prioridad', 'Media')),
            'prioridad_display': data.get('prioridad', 'Media'),
            'assigned_to': data.get('assigned_to'),
            'assigned_to_name': data.get('assigned_to_name', ''),
            'fecha_reporte': fecha_reporte,
            'fecha_asignacion': self._parse_timestamp(data.get('assigned_at')),
            'reportado_por': data.get('user_id', 'Desconocido'),
            'version_app': data.get('device_info', ''),

            # AI Classification fields (additional info)
            'is_ai_classified': data.get('is_ai_classified', False),
            'ai_confidence': data.get('ai_confidence', 0),
            'ai_suggested_classification': data.get('ai_suggested_classification', ''),
            'ai_model_version': data.get('ai_model_version', ''),

            # Raw Firestore data (for debugging)
            '_firestore_data': data,
        }

    def _map_clasificacion_to_django(self, firestore_value):
        """
        Map Firestore classification to Django model choices

        Args:
            firestore_value: Classification from Firestore

        Returns:
            Django-compatible classification value
        """
        # Normalize to lowercase
        value = str(firestore_value).lower().strip()

        # Direct mapping
        mapping = {
            'orgánico': 'organico',
            'organico': 'organico',
            'plástico': 'plastico',
            'plastico': 'plastico',
            'vidrio': 'vidrio',
            'papel': 'papel',
            'papel/cartón': 'papel',
            'cartón': 'papel',
            'metal': 'metal',
            'electrónico': 'electronico',
            'electronico': 'electronico',
            'textil': 'textil',
            'peligroso': 'peligroso',
            'construcción': 'construccion',
            'construccion': 'construccion',
        }

        return mapping.get(value, 'otros')

    def _map_estado_to_django(self, firestore_estado):
        """
        Map Firestore status to Django model status choices

        Args:
            firestore_estado: Status from Firestore

        Returns:
            Django-compatible status value
        """
        mapping = {
            'received': 'pendiente',
            'recibido': 'pendiente',
            'pendiente': 'pendiente',
            'assigned': 'asignado',
            'asignado': 'asignado',
            'in_progress': 'en_proceso',
            'en_proceso': 'en_proceso',
            'completed': 'resuelto',
            'resuelto': 'resuelto',
            'finalizado': 'resuelto',
            'cancelado': 'cancelado',
            'cancelled': 'cancelado',
        }

        value = str(firestore_estado).lower().strip()
        return mapping.get(value, 'pendiente')

    def _map_prioridad_to_django(self, firestore_prioridad):
        """
        Map Firestore priority to Django model priority choices

        Args:
            firestore_prioridad: Priority from Firestore

        Returns:
            Django-compatible priority value
        """
        mapping = {
            'baja': 'baja',
            'low': 'baja',
            'media': 'media',
            'medium': 'media',
            'alta': 'alta',
            'high': 'alta',
            'urgente': 'urgente',
            'urgent': 'urgente',
        }

        value = str(firestore_prioridad).lower().strip()
        return mapping.get(value, 'media')

    def _parse_timestamp(self, timestamp):
        """
        Parse Firestore timestamp to Python datetime

        Args:
            timestamp: Firestore timestamp or None

        Returns:
            Python datetime or None
        """
        if timestamp and hasattr(timestamp, 'timestamp'):
            return datetime.fromtimestamp(timestamp.timestamp())
        return None


# Singleton instance
firestore_service = FirestoreService()
